Add support for testing whether a value is a constant-cased string

Pranavchiku opened the pull request:
Resolves #534.

## Checklist

- [x] Add readme.md
- [x] Add package.json
- [x] Add lib
- [x] Add examples
- [x] Add docs
- [x] Add etc
- [x] Add bin
- [x] Add benchmark
- [x] Add test

* * *

@kgryte 

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Pranavchiku committed:
Add readme.md

Pranavchiku commented:
I don't have the access to add assignees for the PR :/

Edit: Thanks :)

kgryte commented:
@Pranavchiku Sorry about that. This is the first time I'm trying this out. Just assigned. Let's see what happens.

kgryte commented:
Looks like the workflow is stuck in a queue, so we'll need to wait a while for it to kick off. How long we'll have to wait is TBD.

Pranavchiku commented:
>Looks like the workflow is stuck in a queue, so we'll need to wait a while for it to kick off. How long we'll have to wait is TBD.

It might get failed if stuck there for long time ( I faced this issue once at lfortran )

Planeshifter committed:
Add README.md

Planeshifter commented:
Okay, so there are currently problems with the AI scaffolding workflow when triggered from a fork, which we will need to look into.  Apologies!

As a temporary workaround, I have created a `assert/is-constantcase` branch with your README.md and the files: https://github.com/stdlib-js/stdlib/pull/541 

Feel free to either checkout that branch or copy the files over into this PR's branch.

Pranavchiku commented:
I rebased it over my PR, and added implementation, made a clean build again, there are quite a few bugs like output not correctly predicted, linting errors, and a few things will be needed to fix before I can push those changes in this PR, will try to get that done as soon as possible :)

kgryte commented:
Thanks for trying this out @Pranavchiku!

Pranavchiku commented:
Sorry for delaying it, will try to get the things up as soon as possible.

Pranavchiku committed:
add package.json

Pranavchiku committed:
Add lib

Pranavchiku committed:
Add examples

Pranavchiku committed:
Add etc

Pranavchiku committed:
Add docs

Pranavchiku committed:
Add bin

Pranavchiku committed:
add benchmark

Pranavchiku commented:
so,
```javascript
tape( 'the command-line interface supports use as a standard stream', opts, function test( t ) {
	var cmd = [
		'printf "BEEP\nboop"',
		'|',
		EXEC_PATH,
		fpath
	];

	exec( cmd.join( ' ' ), done );

	function done( error, stdout, stderr ) {
		if ( error ) {
			t.fail( error.message );
		} else {
			t.strictEqual( stdout.toString(), 'true\ntrue\n', 'expected value' );
			t.strictEqual( stderr.toString(), '', 'does not print to `stderr`' );
		}
		t.end();
	}
});
```
This test is not passing in ```test.cli.js```, idk the reason, pasting the logs below
```
  the command-line interface supports use as a standard stream


    ✖ expected value
    -----------------
      operator: equal
      expected: 'false\nfalse\n'
      actual:   'false\n'
      at: done (/home/pranavchiku/stdlib/lib/node_modules/@stdlib/assert/is-constantcase/test/test.cli.js:179:6)

    ✔ does not print to `stderr`
    ```

kgryte commented on lib/node_modules/@stdlib/assert/is-constantcase/bin/cli:
```diff
@@ -0,0 +1,103 @@
+		stdin( onRead );
+		return;
+	}
+	if ( cli.args.length ) {
+		console.log( isConstantCase( cli.args[ 0 ] ) ); // eslint-disable-line no-console
+	} else {
+		console.log( isConstantCase() ); // eslint-disable-line no-console
+	}
+
+	/**
+	* Callback invoked upon reading from `stdin`.
+	*
+	* @private
+	* @param {(Error|null)} error - error object
+	* @param {Buffer} data - data
+	* @returns {void}
+	*/
+	function onRead( error, data ) {
+		var str;
+		var i;
+		if ( error ) {
+			return cli.error( error );
+		}
+		str = data.toString();
+		if ( opts.sep ) {
+			str = str.split( opts.sep );
+			for ( i = 0; i < str.length; i++ ) {
+				console.log( isConstantCase( str[ i ] ) ); // eslint-disable-line no-console
+			}
+		} else {
+			console.log( isConstantCase( str ) ); // eslint-disable-line no-console
+		}
```
This logic is incorrect. See https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/assert/is-kebabcase/bin/cli#L80 for the correct logic.

cc @Planeshifter

Pranavchiku replied:
oh okay, I actually will need to rewrite the tests.

kgryte replied:
The tests should be fine, unless you changed them. It is just the logic in the CLI file which needs updating.

Pranavchiku replied:
oh okay, I was not able to push those changes after I rebase it over my current PR, it threw me error, hence I had to change those.

kgryte commented:
@Pranavchiku Left a comment. You may need to merge in changes from upstream. Currently, this PR is a bit "dirty" in that it contains changes/files which are already present in the upstream repo.

Pranavchiku committed:
Add test.js

Pranavchiku commented:
Yeah I will do that.

Pranavchiku commented:
I observed few things on the files created by ```stdlib-bot```

1. Ending the file with two extra lines
2. Using extra line after `use-strict`
3. Tests logic are not great
4. ```test/fixtures``` file is not created

Rest everything was very perfect!

kgryte commented:
@Pranavchiku That's great feedback! Thanks! cc @Planeshifter

Pranavchiku committed:
add test/fixtures

Pranavchiku committed:
Add test.cli.js

Pranavchiku commented:
I am currently getting a small error in ```test.cli.js```, I will resolve that, polish the PR and then push changes as soon as possible :)

kgryte commented:
@Pranavchiku Awesome! Thanks for working on this!

Pranavchiku committed:
add test.cli.js

Pranavchiku commented:
Okay so in here ```the command-line interface supports specifying a custom delimiter when used as a standard stream (regexp)``` is not supported, I don't know the reason why it kept failing, if you can have a check

Pranavchiku committed:
update bin/cli

kgryte commented on lib/node_modules/@stdlib/assert/is-constantcase/test/test.cli.js:
```diff
@@ -0,0 +1,241 @@
+		}
+		t.end();
+	}
+});
```
Missing tests for when the custom delimiter is a regexp. Can copy over from `@stdlib/assert/is-kebabcase` or similar package.

Pranavchiku committed:
add test.cli.js

kgryte commented on lib/node_modules/@stdlib/assert/is-constantcase/test/test.cli.js:
```diff
@@ -164,7 +164,7 @@ tape( 'the command-line interface tests if a string argument contains only upper
 tape( 'the command-line interface supports use as a standard stream', opts, function test( t ) {
 	var cmd = [
-		'printf "Beepboop"',
+		'printf "Beepboop\nbeepBoop"',
```
May be better to have both a positive and a negative test value here. E.g.,

```
Beepboop\nBEEPBOOP
```

kgryte commented on lib/node_modules/@stdlib/assert/is-constantcase/test/test.cli.js:
```diff
@@ -205,6 +205,28 @@ tape( 'the command-line interface supports specifying a custom delimiter when us
 
+tape( 'the command-line interface supports specifying a custom delimiter when used as a standard stream (regexp)', opts, function test( t ) {
+	var cmd = [
+		'printf \'beep boop\tcheckHere\'',
```
Same comment as above.

Pranavchiku committed:
performed suggested changes

Pranavchiku commented:
@kgryte you can have a final review.

kgryte commented on lib/node_modules/@stdlib/assert/is-constantcase/README.md:
````diff
@@ -0,0 +1,202 @@
+By default, when used as a [standard stream][standard-streams], the implementation assumes newline-delimited data. To specify an alternative delimiter, set the `split` option.
+
+```bash
+$ echo -n 'beepBoop\tbeep_boop' | is-constantcase --split '\t'
````
```suggestion
$ echo -n 'beepBoop\tBEEP_BOOP' | is-constantcase --split '\t'
```

kgryte committed:
Fix example

kgryte commented on lib/node_modules/@stdlib/assert/is-constantcase/README.md:
````diff
@@ -0,0 +1,202 @@
+### Examples
+
+```bash
+$ is-constantcase beepboop
````
```suggestion
$ is-constantcase BEEP_BOOP
```

kgryte committed:
Fix example

kgryte commented on lib/node_modules/@stdlib/assert/is-constantcase/docs/repl.txt:
```diff
@@ -0,0 +1,23 @@
+    Returns
+    -------
+    bool: boolean
+        Boolean indicating whether value is a constantcase string.
```
```suggestion
        Boolean indicating whether a value is a constantcase string.
```

kgryte committed:
Update description

kgryte commented on lib/node_modules/@stdlib/assert/is-constantcase/docs/types/index.d.ts:
```diff
@@ -0,0 +1,54 @@
+// TypeScript Version: 2.0
+
+/**
+* Tests if a `value` is a constantcase `string`.
+*
+* @param value - value to test
+* @returns boolean indicating whether a `value` is a constantcase `string`
```
```suggestion
* Tests if a value is a constantcase string.
*
* @param value - value to test
* @returns boolean indicating whether a value is a constantcase string
```

kgryte committed:
Update descriptions

kgryte commented on lib/node_modules/@stdlib/assert/is-constantcase/lib/main.js:
```diff
@@ -0,0 +1,67 @@
+// MAIN //
+
+/**
+* Tests if a `value` is a constantcase `string`.
+*
+* @param {*} value - value to test
+* @returns {boolean} boolean indicating whether a `value` is a constantcase `string`
```
```suggestion
* Tests if a value is a constantcase string.
*
* @param {*} value - value to test
* @returns {boolean} boolean indicating whether a value is a constantcase string
```

kgryte committed:
Update descriptions

kgryte reviewed (approved):
LGTM. Thanks, @Pranavchiku!

Planeshifter commented:
> I observed few things on the files created by `stdlib-bot`
> 
> 1. Ending the file with two extra lines
> 2. Using extra line after `use-strict`
> 3. Tests logic are not great
> 4. `test/fixtures` file is not created
> 
> Rest everything was very perfect!

Thank you @Pranavchiku for the feedback. 1., 2., and 4. will be fixed in the next iteration of the scaffolding tool. We will keep iterating to hopefully get to a very smooth development experience in the near future.
